File: haskell/0001.hs
-- Multiples of 3 or 5:
-- If we list all the natural numbers below 10 that
-- are multiples of 3 or 5, we get 3, 5, 6 and 9.
-- The sum of these multiples is 23.
-- Find the sum of all the multiples of 3 or 5 below 1000.

divByThreeOrFive :: Integer -> Bool
divByThreeOrFive n = (mod n 3) == 0 || (mod n 5) == 0

sequ :: (Integer -> Bool) -> Integer -> [Integer]
sequ func start =
    if func start then start : (sequ func (start + 1)) else sequ func (start + 1)

result :: Integer
result = foldl (+) 0 (takeWhile (< 1000) (sequ divByThreeOrFive 0))

-- ghci
-- :load 0001.hs
-- divByThreeOrFive 3
-- divByThreeOrFive 4
-- divByThreeOrFive 5
-- takeWhile (< 1000) (sequ divByThreeOrFive 0)
-- foldl (+) 0 (takeWhile (< 1000) (sequ divByThreeOrFive 0))
-- 233168

File: prolog/0001.pl
% Multiples of 3 or 5:
% If we list all the natural numbers below 10 that
% are multiples of 3 or 5, we get 3, 5, 6 and 9.
% The sum of these multiples is 23.
% Find the sum of all the multiples of 3 or 5 below 1000.

:- include('util.pl').

is_multiple(I) :-
    0 =:= I mod 3;
    0 =:= I mod 5.

start(SumMultiples, Limit) :-
    Limit0 #= Limit - 1,
    do_list(Limit0, List),
    include(is_multiple, List, ListMultiple),
    sum_list(ListMultiple, SumMultiples).

% gprolog
% ['0001'].
% start(SumMultiples, 10).
% start(SumMultiples, 1000).
% 233168
